Recovery method not found for method in class that implements an interface

I have a component class that implements an interface. The interface is being autowired into my calling class, and in my implementation class I have annotated a method from the interface with `@Retryable` and added a public recovery method. However, when the retry fails, it fails to find the recovery method.

Through some debugging, I have observed that `ReflectionUtils#doWithMethods(Class, MethodCallback)` is used to iterate through the methods of the failing method's class to find the recovery method. However, in my case, `failingMethod.getDeclaringClass()` is returning my interface instead of my implementation class, and because the interface doesn't contain my recovery method, the recovery method is never found. I noticed that the object of my implementation class is passed to the `RecoverAnnotationRecoveryHandler#init(Object, Method)` method in the target parameter, but never used.

```
	private void init(Object target, Method method) {
		final Map<Class<? extends Throwable>, Method> types = new HashMap<Class<? extends Throwable>, Method>();
		final Method failingMethod = method;
		ReflectionUtils.doWithMethods(failingMethod.getDeclaringClass(),
```

Is there a reason that `failingMethod.getDeclaringClass()` is being used instead of `target.getClass()`? Thank you for looking at this issue. Let me know if I should open a PR to attempt to fix this.

Currently using Spring Retry 1.2.5.RELEASE with Spring Boot 2.2.5.RELEASE.

@dsyer There already is a test here 

https://github.com/spring-projects/spring-retry/blob/4fe92bb20f4beb158b2934a34a6aa0fa1c10df8e/src/test/java/org/springframework/retry/annotation/EnableRetryTests.java#L172-L181

Which fails at line 179 when running in an IDE (with your fix backed out).

However maven reports

```
[INFO] ------------------------------------------------------------------------
[INFO] BUILD SUCCESS
[INFO] ------------------------------------------------------------------------
```

(Which is why I missed it while rebasing the PR).

But if I look further up in the log...

```
INFO] 
[INFO] Results:
[INFO] 
[ERROR] Errors: 
[ERROR]   EnableRetryTests.testInterface:179 » ExhaustedRetry Cannot locate recovery met...
[INFO] 
```

--- a/src/main/java/org/springframework/retry/annotation/RecoverAnnotationRecoveryHandler.java
+++ b/src/main/java/org/springframework/retry/annotation/RecoverAnnotationRecoveryHandler.java
@@ -166,7 +166,7 @@ public class RecoverAnnotationRecoveryHandler<T> implements MethodInvocationReco
 		if (retryable != null) {
 			this.recoverMethodName = retryable.recover();
 		}
-		ReflectionUtils.doWithMethods(failingMethod.getDeclaringClass(), new MethodCallback() {
+		ReflectionUtils.doWithMethods(target.getClass(), new MethodCallback() {
 			@Override
 			public void doWith(Method method) throws IllegalArgumentException, IllegalAccessException {
 				Recover recover = AnnotationUtils.findAnnotation(method, Recover.class);

--- a/src/test/java/org/springframework/retry/annotation/EnableRetryTests.java
+++ b/src/test/java/org/springframework/retry/annotation/EnableRetryTests.java
@@ -182,6 +182,14 @@ public class EnableRetryTests {
 	}
 
 	@Test
+	public void testInterfaceWithNoRecover() {
+		AnnotationConfigApplicationContext context = new AnnotationConfigApplicationContext(TestConfiguration.class);
+		NoRecoverInterface service = context.getBean(NoRecoverInterface.class);
+		service.service();
+		assertTrue(service.isRecovered());
+	}
+
+	@Test
 	public void testImplementation() {
 		AnnotationConfigApplicationContext context = new AnnotationConfigApplicationContext(TestConfiguration.class);
 		NotAnnotatedInterface service = context.getBean(NotAnnotatedInterface.class);
@@ -349,6 +357,11 @@ public class EnableRetryTests {
 		@Bean
 		public TheInterface anInterface() {
 			return new TheClass();
+		}
+
+		@Bean
+		public NoRecoverInterface anInterfaceWithNoRecover() {
+			return new NoRecoverClass();
 		}
 
 		@Bean
@@ -635,6 +648,32 @@ public class EnableRetryTests {
 			return this.recovered;
 		}
 
+	}
+
+	public static interface NoRecoverInterface {
+		void service();
+		boolean isRecovered();
+	}
+
+	public static class NoRecoverClass implements NoRecoverInterface {
+
+		private boolean recovered;
+
+		@Override
+		@Retryable
+		public void service() {
+			throw new RuntimeException("Planned");
+		}
+
+		@Recover
+		public void recover(Exception e) {
+			this.recovered = true;
+		}
+
+		@Override
+		public boolean isRecovered() {
+			return this.recovered;
+		}
 	}
 
 	public static interface NotAnnotatedInterface {

